File: BOJ_1003/BOJ_1003.cpp
#include <iostream>

using namespace std;
int RECOLLECT[50];

int FIBONACCI(int RECALL);


int main(){
        int N, CASE;
        cin>> CASE;
        for(int i; i<CASE; i++){
                cin>>N;
                if(N == 0){
                        cout<<"1 0\n";
                }
                else if(N == 1){
                        cout<<"0 1\n";
                }
                else if(N == 2){
                        cout<<"1 1\n";
                }
                else{
                        FIBONACCI(N);
                        cout<<RECOLLECT[N-1]<<" "<<RECOLLECT[N]<<endl;
                }
        }

        return 0;
}
int FIBONACCI(int RECALL){
        if(RECALL <= 0){
                RECOLLECT[0] = 0;
                return 0;
        }
        else if(RECALL == 1){
                RECOLLECT[1] = 1;
                return 1;
        }
        if(RECOLLECT[RECALL] != 0){
                return RECOLLECT[RECALL];
        }
        else{
                RECOLLECT[RECALL] = FIBONACCI(RECALL-1) + FIBONACCI(RECALL-2);
                return RECOLLECT[RECALL];
        }
}

File: BOJ_1003/BOJ_1003.cs
using System;

namespace FIBONACCI{
        public class FIBONACCI{
                static void Main(){
                        int CASE = Convert.ToInt32(Console.ReadLine());
                        for(int i=0; i<CASE; i++){
                                int N = Convert.ToInt32(Console.ReadLine());
                                if(N == 0){
                                        Console.WriteLine("1 0");
                                }
                                else if(N == 1){
                                        Console.WriteLine("0 1");
                                }
                                else if(N == 2){
                                        Console.WriteLine("1 1");
                                }
                                else{
                                        FIBOS(N);
                                        Console.WriteLine("{0} {1}", TABLE[N-1], TABLE[N]);
                                }
                        }


                }
                public static int[] TABLE = new int[41];
                static int FIBOS(int RECALL){
                        if(RECALL <= 0){
                                TABLE[0] = 0;
                                return 0;
                        }
                        else if(RECALL == 1){
                                TABLE[1] = 1;
                                return 1;
                        }
                        if(TABLE[RECALL] != 0){
                                return TABLE[RECALL];
                        }
                        else{
                                TABLE[RECALL] = FIBOS(RECALL-1) + FIBOS(RECALL-2);
                                return TABLE[RECALL];
                        }
                }
        }
}
